src/paperoni/refinement/fetch.py

```diff
@@ -8,26 +11,78 @@ def fetch(type: str, link: str):
     return None
 
 
-def register_fetch(f):
-    f.description = f.__name__
-    fetch.register(f)
-    return f
+def register_fetch(f=None, *, tags=None):
+    def decorator(f):
+        f.description = f.__name__
+        f.tags = tags or set()
+        fetch.register(f)
+        return f
+
+    if f is None:
+        return decorator
+    else:
+        return decorator(f)
+
+
+def _test_tags(f_tags: set, tags: set) -> bool:
+    # no tags are required for the function to execute
+    if not f_tags:
+        return True
+
+    try:
+        tags.remove("*")
+        star_tag = "*"
+    except KeyError:
+        star_tag = None
+
+    # tags that are not in f_tags
+    extra_tags = tags - (f_tags & tags)
+
+    if extra_tags:
+        return False
 
+    unmatched_f_tags = f_tags - tags
 
-def fetch_all(type, link, statuses=None):
+    # There should be no unmatched tags, or if there are, there should be a star
+    # tag
+    return not unmatched_f_tags or star_tag is not None
+
+
+def _call(f: Callable, *args, force: bool = False, **kwargs) -> tuple:
+    f_sig = inspect.signature(f)
+    if "force" in f_sig.parameters:
+        ret = f(*args, force=force, **kwargs)
+    else:
+        ret = f(*args, **kwargs)
+
+    if isinstance(ret, tuple):
+        return ret
+    else:
+        return (ret,)
+
+
+def fetch_all(type, link, statuses=None, tags=None, force=False):
     statuses = statuses or {}
+    tags = tags or set()
     for f in fetch.resolve_all(type, link):
+        if not _test_tags(getattr(f.func, "tags", set()), tags):
+            continue
+
         name = getattr(f.func, "description", "???")
         key = f"{type}:{link}"
         nk = name, key
         if nk in statuses:
             continue
         statuses[nk] = "pending"
         try:
-            paper = f()
+            paper, *subnames = _call(f.func, type, link, force=force)
             if paper is not None:
                 statuses[nk] = "found"
-                yield PaperInfo(paper=paper, key=key, info={"refined_by": {name: key}})
+                yield PaperInfo(
+                    paper=paper,
+                    key=key,
+                    info={"refined_by": {":".join([name, *subnames]): key}},
```

Comment:
The value for refined_by, ideally, should not depend on the output of the call. That's because we want to be able to pre-populate the statuses dict with all the refinements we've already done, in order to skip them (i.e. *not* call the function).

Reply:
fixed